host/storage/storage.go

```diff
@@ -321,6 +325,49 @@ func (vm *VolumeManager) shrinkVolume(ctx context.Context, id int64, volume *vol
 	return nil
 }
 
+// writeSector atomically adds a sector to the database and writes it to disk
+func (vm *VolumeManager) writeSector(root types.Hash256, data *[proto4.SectorSize]byte) error {
+	return vm.vs.StoreSector(root, func(loc SectorLocation) error {
```

Comment:
I have a slightly unrelated question regarding the concurrency within `StoreSector`.

From what I can tell we have 2 db transactions and in-between those we write the sector to a location.
If that fails, the sector is freed again.

But at a first glance it seems like there might be a race when uploading the same sector twice at the same time when the first write fails. e.g. the first write allocates the empty location, is then interrupted, then the second write fetches the same location and successfully writes the sector to disk. Then the first write resumes, fails for some reason and ends up freeing the location that the other thread just successfully wrote to in the rollback.

Obviously that is incredibly unlikely since it requires very particular timing, an odd use-case and strange disk behavior, but I wanted to highlight it regardless.